Java Constant Converter

kennyqiyenkan opened the pull request:
Users can now use the `constants` to add constants to their converted java code.

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into Java-ConstantConverter

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into Java-ConstantConverter

kennyqiyenkan committed:
Added `constants` to java converter

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into Java-ConstantConverter

kennyqiyenkan committed:
formatting

kennyqiyenkan committed:
formatting

again.

kgrgreer commented on src/foam/java/Class.js:
```diff
@@ -100,6 +112,11 @@ foam.CLASS({
     },
 
+    function constant(c) {
+      this.constants.push(foam.java.constant.create(c));
```
Should this be foam.java.Constant? How is this working?

kennyqiyenkan committed:
Changed constant to proper Constant

kgrgreer commented on src/foam/java/Constant.js:
```diff
@@ -0,0 +1,72 @@
+
+  properties: [
+    'name',
+    'javaType',
```
This is already java specific, so we just need 'type' with java being implied.
Also, we only need 'value', which is the 'javaValue'.

kennyqiyenkan replied:
type is understandable.

as for javaValue/value, how would we be able to discern if they have their own custom instantiation or they are simply using a primitive?

kgrgreer commented on src/foam/java/Constant.js:
```diff
@@ -0,0 +1,72 @@
+
+      if ( this.documentation != undefined ) {
+        o.out('/* ', this.documentation, ' */\n');
+        o.indent();
```
What are the next two lines for?

kennyqiyenkan replied:
the tabbing doesn't come out well after the `\n`. The two indents fixes that.

kgrgreer commented on src/foam/java/Constant.js:
```diff
@@ -0,0 +1,72 @@
+
+      o.out('public static final ');
+
+      if ( this.type == undefined ) {
```
This logic should be in the factory for the 'type' property.

kennyqiyenkan replied:
what would you mean by that? As in we do not need to check if the type is undefined?

kgrgreer commented on src/foam/java/Constant.js:
```diff
@@ -0,0 +1,72 @@
+      // but not declare a javaType for it, the Java code should break on compile
+      if ( this.javaValue == undefined ) {
+        if ( foam.String.isInstance(this.value) ) {
+          o.out('\"' + this.value + '\"');
```
You'll need to escape this.value in case it contains "'s.

kennyqiyenkan replied:
Should that not be handled by the user?

kgrgreer replied:
The user could do something like:
constants: { QUOTE_CHARACTER: '"' } OR
constants: { QUOTE_CHARACTER: "\"" }

Which you would output as """. How could the user handle this?

kennyqiyenkan replied:
mm.. I believe I see your point.

kennyqiyenkan committed:
Changed code based on comments

- removed `javaType` and only use `type`
- escaped incoming value
- check for negative int

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into Java-ConstantConverter

kennyqiyenkan committed:
Fix logic error

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into Java-ConstantConverter
